Hoist HTTP match types into type-specific structs.

jpeach opened the pull request:
Update the route match struct to hoist the path and header match
specifications into their own structs. This improves API clarity and
consistency by using the same matching syntax for paths and headers.

Update the matching defaults at various levels so that there is always a
default prefix path match on "/". This reduces the amount of YAML needed
for simple cases and removes the validation burden from controllers.

Signed-off-by: James Peach <jpeach@vmware.com>

jpeach commented:
This PR still needs polish around doc comments. Posting for WG discussion and feedback.

xref #241

robscott commented:
/ok-to-test

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -115,9 +115,8 @@ const (
-// HTTPRouteMatch defines the predicate used to match requests to a
-// given action.
-type HTTPRouteMatch struct {
+type HTTPPathMatch struct {
```
Although I understand the desire to have path match be consistent with header match, I think this ends up making a more significant change than we want. This change results in path match becoming optional where before it was defaulted to match `/`. It also results in an ambiguous situation where nothing is actually required or set by default in `HTTPRouteMatch`. To avoid that, I'd lean towards only changing header matching with this PR.

jpeach replied:
Yep, we want to be careful about the semantics here. I'm OK with dropping the '/' path match because implicitly requiring an exact match on '/' seems odd now that other match types are available. Being able to produce an empty HTTPRouteMatch is weird; you could argue equally that it should match everything or nothing.

robscott replied:
Yeah I completely agree here, but had a slightly different idea for a solution. I filed a [follow up issue](https://github.com/kubernetes-sigs/service-apis/issues/255) that suggests we should be defaulting to a `Prefix` match here so the default `/` would match all paths.

yiyangy replied:
+1 on Rob's proposal of leaving path match in the `HTTPRouteMatch` struct and defaulting to `'type: Prefix, path: "/"'`.  It would make the spec with default match simpler and less ambiguous.

hbagdi replied:
I've been wondering for a few weeks and I can't recall the reason of why we put in `Exact` match as the default for `PathMatchType` because it seems draconian. Anyone recall the reason?

I agree with Rob and Yi here. Header based routing in addition to path based is more common in practice and less error-prone.

jpeach replied:
> Although I understand the desire to have path match be consistent with header match, I think this ends up making a more significant change than we want. 

I mulled this over and I think that we should make these consistent. Maning the API consistent and predictable is really important, and having different syntaxes for different kind of match actually adds complexity rather than removing it.


This is what you have to do now, which is inconsistent, a bit ugly (`patchMatchType`), and really needs handy reference to the docs to write:
```yaml
   - match:
         path: /bar
         pathMatchType: Prefix
         headers:
           type: Contains
           values:
              foo: bar
```

jpeach replied:
Resolving as per WG discussion and consensus in favor of this.

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
 	// +kubebuilder:validation:Enum=Exact;ImplementationSpecific
 	// +kubebuilder:default=Exact
-	HeaderMatchType *string `json:"headerMatchType" protobuf:"bytes,3,opt,name=headerMatchType"`
+	Type *string `json:"type" protobuf:"bytes,3,opt,name=type"`
```
Does this need to be a pointer? It seems like nil and "" are indistinguishable here and we're already setting a default value.

yiyangy replied:
I feel this no longer needs to be a pointer. If I recall correctly, having this field be non-pointer and default to `Exact` but avoiding the case where no header match is meant to be set was the initial motivation to make `HTTPHeaderMatch` into its own struct.

jpeach replied:
Yeh, we can remove the pointer, since the defaulting has precedence anyway.

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
+
 	// +optional
-	Headers map[string]string `json:"headers" protobuf:"bytes,4,rep,name=headers"`
+	Header *HTTPHeaderMatch `json:"headers" protobuf:"bytes,7,opt,name=headers"`
```
I think we should have a comment describing the field above this.

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
 	// +kubebuilder:default=Exact
-	HeaderMatchType *string `json:"headerMatchType" protobuf:"bytes,3,opt,name=headerMatchType"`
+	Type *string `json:"type" protobuf:"bytes,3,opt,name=type"`
 	// Headers are the HTTP Headers to match as interpreted via
```
Nit: I think this should be "Values" now to match the field name.

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
-	HeaderMatchType *string `json:"headerMatchType" protobuf:"bytes,3,opt,name=headerMatchType"`
+	Type *string `json:"type" protobuf:"bytes,3,opt,name=type"`
 	// Headers are the HTTP Headers to match as interpreted via
 	// HeaderMatchType. Multiple headers are ANDed together, meaning, a request
```
Nit: I think HeaderMatchType should also be updated now to reflect the new simplified type.

yiyangy commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
+
 	// +optional
-	Headers map[string]string `json:"headers" protobuf:"bytes,4,rep,name=headers"`
+	Header *HTTPHeaderMatch `json:"headers" protobuf:"bytes,7,opt,name=headers"`
```
As we are making `HTTPHeaderMatch` into its own struct, are there use cases where a `HTTPRouteMatch` includes multiple types of `HTTPHeaderMatch`? i.e. Does it make sense making this into a list? For example,
```
Headers []HTTPHeaderMatch
```
While `Values` in `HTTPHeaderMatch` struct can be updated to be a single value or remain a list.

jpeach replied:
Yeh, this is definitely worth trying out.

jpeach replied:
I think that for now the right approach is to land #241, and then specify multiple matches for this case. It will be equivalent (though more verbose) as well as more verbose. WDYT?

yiyangy commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -145,13 +146,23 @@ type HTTPRouteMatch struct {
+
 	// +optional
-	Headers map[string]string `json:"headers" protobuf:"bytes,4,rep,name=headers"`
+	Header *HTTPHeaderMatch `json:"headers" protobuf:"bytes,7,opt,name=headers"`
```
s/headers/header

jpeach replied:
I could go either way on this. To me, `headers` reads better because people colloquially talk about "HTTP Headers", but I could easily be in the minority on this one :)

jpeach commented:
From WG discussion; we should add kubebuilder defaulting annotations such that not specifying any matches ends up doing a prefix path match on '/'. This means that a route rule action with no match field would by default match everything.

jpeach commented:
> From WG discussion; we should add kubebuilder defaulting annotations such that not specifying any matches ends up doing a prefix path match on '/'. This means that a route rule action with no match field would by default match everything.

Rebased and added the defaulting, which works great!

robscott commented on Makefile:
```diff
@@ -27,13 +27,16 @@ all: generate controller verify
 	$(MAKE) -f kubebuilder.mk generate
 	$(MAKE) manifests
 	$(MAKE) docs
+	$(MAKE) proto
```
This seems like it might be out of scope for this PR?

jpeach replied:
I can split this into a separate PR if you prefer?

jpeach replied:
Moved into #267

robscott commented on apis/v1alpha1/httproute_types.go:
````diff
@@ -86,19 +86,35 @@ type HTTPRouteRule struct {
+	//
 	// For example, take the following matches configuration:
+	//
+	// ```yaml
````
This feels strange to see in go comments, does this end up helping the reference doc?

jpeach replied:
Unfortunately not. I haven't been able to get the API doc generator to emit reformatted text of any kind. I can just remove this, since it doesn't really help.

jpeach replied:
I removed the 'yaml', but kept the markdown formatting so that it formats as a block in the API docs (indentation is still wrong though).

robscott commented on apis/v1alpha1/httproute_types.go:
```diff
@@ -183,14 +193,50 @@ type HTTPRouteMatch struct {
+	// names MUST be matched case-insensitively.
+	//
+	// Multiple match values are ANDed together, meaning, a request
+	// must match all the specified header to select the route.
```
Nit: headers

jpeach replied:
fixed.

jpeach commented on examples/default-match-http.yaml:
```diff
@@ -0,0 +1,73 @@
+    rules:
+    - matches:
+      - headers:
+          type: Exact
```
`matchType` is a bit verbose for my taste, but `match` works pretty well:

```yaml
spec:
  hosts:
  - hostnames:
    - default-match.com
    rules:
    - matches:
      - headers:
          match: Exact
          values:
            magic: default-match
      action:
        forwardTo:
          - targetRef:
              name: my-service
              group: core
              resource: services
    - matches:
      - path:
          match: Exact
          value: /example/exact
      action:
        forwardTo:
          - targetRef:
              name: my-service
              group: core
              resource: services
```

hbagdi replied:
I'd prefer `type` over `match`. `match` already is used in a very close by context and we should avoid overloading it.

hbagdi commented:
@jpeach Needs a rebase.

/assign @robscott

robscott commented:
Thanks for the work on this @jpeach! This LGTM, just needs another rebase.

/lgtm

hbagdi commented:
Rebase checks out.
/lgtm
/assign @bowei

bowei commented:
/lgtm
/approve
